File: web/src/features/proof/retrievedContext.ts
// Heuristic parser for the "retrieved public context" shape some bench datasets flatten into a
// single input_text (the Advisor governance corpus convention): a Question, then a
// "Retrieved public context:" block of repeating `Source N: <id>` records, each optionally carrying
// Label / Class / Title / Excerpt. This is a CONVENTION, not a universal schema — an arbitrary
// imported JSONL set has free-form input_text — so the parser is detect-and-degrade: it returns the
// structured form ONLY when the markers are unmistakable, and `null` otherwise so the caller falls
// back to rendering the raw field. Pure + unit-tested; the React layer stays dumb.

export interface RetrievedSource {
  id: string;
  label?: string;
  class?: string;
  title?: string;
  excerpt?: string;
}

export interface RetrievedContext {
  question: string;
  sources: RetrievedSource[];
}

const CONTEXT_MARKER = "Retrieved public context:";
// A source block opens with `Source <n>: <id>` on its own line.
const SOURCE_LINE = /^Source\s+\d+:\s*(.+)$/;
// The known sub-fields, mapped to the record key they populate. Unknown `Key: value` lines inside a
// block are ignored (they fold into the currently-open field as continuation text instead).
const FIELD_KEYS: Record<string, keyof Omit<RetrievedSource, "id">> = {
  Label: "label",
  Class: "class",
  Title: "title",
  Excerpt: "excerpt",
};
const FIELD_LINE = /^(Label|Class|Title|Excerpt):\s*(.*)$/;

// Parse the flattened retrieved-context shape, or return null when it isn't present. Returning null
// (not a degenerate object) is the contract the caller relies on to fall back to the plain field.
export function parseRetrievedContext(input: string): RetrievedContext | null {
  if (!input) return null;
  const markerAt = input.indexOf(CONTEXT_MARKER);
  if (markerAt === -1) return null;

  // The question is everything before the marker, with a leading "Question:" label stripped.
  const head = input.slice(0, markerAt).trim();
  const question = head.replace(/^Question:\s*/i, "").trim();

  const body = input.slice(markerAt + CONTEXT_MARKER.length);
  const lines = body.split("\n");

  const sources: RetrievedSource[] = [];
  let current: RetrievedSource | null = null;
  let openField: keyof Omit<RetrievedSource, "id"> | null = null;

  const closeField = () => {
    // Trim the accumulated field once its block ends (excerpts can run several lines).
    if (current && openField && current[openField] !== undefined) {
      current[openField] = current[openField]!.trim();
    }
  };

  for (const line of lines) {
    const srcMatch = line.match(SOURCE_LINE);
    if (srcMatch) {
      closeField();
      current = { id: srcMatch[1].trim() };
      sources.push(current);
      openField = null;
      continue;
    }
    if (!current) continue; // text before the first Source: line is noise.

    const fieldMatch = line.match(FIELD_LINE);
    if (fieldMatch) {
      closeField();
      const key = FIELD_KEYS[fieldMatch[1]];
      const value = fieldMatch[2];
      current[key] = value;
      openField = value === "" ? null : key;
      continue;
    }

    // A continuation line for the currently-open field (e.g. a multi-line excerpt).
    if (openField) {
      current[openField] = `${current[openField] ?? ""}\n${line}`;
    }
  }
  closeField();

  // Degrade unless we actually found source records — a bare "Retrieved public context:" with no
  // Source blocks is not the structured shape and should render as plain text.
  if (sources.length === 0) return null;
  return { question, sources };
}

File: src/orionfold/receipts/retrieved_context.py
"""Parser for the flattened "retrieved public context" shape that some bench datasets pack into a
single ``input_text`` (the Advisor governance corpus convention): a Question, then a
``Retrieved public context:`` block of repeating ``Source N: <id>`` records, each optionally
carrying Label / Class / Title / Excerpt.

This is a **faithful port** of the cockpit's ``web/src/features/proof/retrievedContext.ts`` so the
receipt renders the SAME structured shape the live cockpit shows for datasets / corpus / evals —
one parsing→rendering solution, two surfaces. Like the TS original it is *detect-and-degrade*: it
returns the structured form ONLY when the markers are unmistakable, and ``None`` otherwise so the
caller falls back to rendering the raw field as-is (an arbitrary imported JSONL set has free-form
``input_text``). Keep the two implementations in sync; the shared test fixtures guard the parity.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

_CONTEXT_MARKER = "Retrieved public context:"
# A source block opens with `Source <n>: <id>` on its own line.
_SOURCE_LINE = re.compile(r"^Source\s+\d+:\s*(.+)$")
# The known sub-fields, mapped to the record attribute they populate. Unknown `Key: value` lines
# inside a block fold into the currently-open field as continuation text instead.
_FIELD_LINE = re.compile(r"^(Label|Class|Title|Excerpt):\s*(.*)$")
_FIELD_ATTR = {"Label": "label", "Class": "class_", "Title": "title", "Excerpt": "excerpt"}


@dataclass
class RetrievedSource:
    id: str
    label: str | None = None
    class_: str | None = None
    title: str | None = None
    excerpt: str | None = None


@dataclass
class RetrievedContext:
    question: str
    sources: list[RetrievedSource]


def parse_retrieved_context(input_text: str) -> RetrievedContext | None:
    """Parse the flattened retrieved-context shape, or return ``None`` when it isn't present.
    Returning ``None`` (not a degenerate object) is the contract the caller relies on to fall back
    to the plain field — mirrors the TS parser exactly."""
    if not input_text:
        return None
    marker_at = input_text.find(_CONTEXT_MARKER)
    if marker_at == -1:
        return None

    # The question is everything before the marker, with a leading "Question:" label stripped.
    head = input_text[:marker_at].strip()
    question = re.sub(r"^Question:\s*", "", head, flags=re.IGNORECASE).strip()

    body = input_text[marker_at + len(_CONTEXT_MARKER) :]
    sources: list[RetrievedSource] = []
    current: RetrievedSource | None = None
    open_field: str | None = None

    def close_field() -> None:
        # Trim the accumulated field once its block ends (excerpts can run several lines).
        if current is not None and open_field is not None:
            val = getattr(current, open_field)
            if val is not None:
                setattr(current, open_field, val.strip())

    for line in body.split("\n"):
        src = _SOURCE_LINE.match(line)
        if src:
            close_field()
            current = RetrievedSource(id=src.group(1).strip())
            sources.append(current)
            open_field = None
            continue
        if current is None:
            continue  # text before the first Source: line is noise.

        field = _FIELD_LINE.match(line)
        if field:
            close_field()
            attr = _FIELD_ATTR[field.group(1)]
            value = field.group(2)
            setattr(current, attr, value)
            open_field = None if value == "" else attr
            continue

        # A continuation line for the currently-open field (e.g. a multi-line excerpt).
        if open_field is not None:
            prev = getattr(current, open_field) or ""
            setattr(current, open_field, f"{prev}\n{line}")

    close_field()

    # Degrade unless we actually found source records — a bare marker with no Source blocks is not
    # the structured shape and should render as plain text.
    if not sources:
        return None
    return RetrievedContext(question=question, sources=sources)
